github/github.go

```diff
@@ -2,121 +2,193 @@ package github
 	if err != nil {
-		return nil, err
+		return "", fmt.Errorf("Error getting access token from github. %v", err)
 	}
-	if resp.StatusCode() > 400 {
-		return nil, fmt.Errorf("github returned: %v", resp.StatusCode())
+	bodyStr := "?" + string(body)
+	parsedUrl, err := url.Parse(bodyStr)
+	parseErr := errors.New("Error parsing github response.")
+	if err != nil {
+		klog.Errorf("Github error parsing response: %v", bodyStr)
+		return "", parseErr
 	}
-	return resp.Body(), nil
+
+	parsedQuery, err := url.ParseQuery(parsedUrl.RawQuery)
+	if err != nil {
+		klog.Errorf("Github error parsing response: %v", bodyStr)
+		return "", parseErr
+	}
+
+	accessToken := parsedQuery["access_token"]
+
+	if len(accessToken) == 0 {
+		klog.Errorf("Github error parsing response: %v", bodyStr)
+		return "", parseErr
+	}
+	return accessToken[0], nil
 }
 
-// A function that returns installation access token (different from JWT token)
-// That token is used in most api calls to identify the org the user belongs to
-// Calls generateJWTToken() and calls a github endpoint to generate the token
-// Note: whenever installation token is used in any endpoint, need to use "token" in
-// "Authorization" header instead of "Bearer"
-func (g *github) GenerateInstallationToken(installationId string) (string, error) {
-	endpoint := fmt.Sprintf("/app/installations/%v/access_tokens", installationId)
+type GithubRepository struct {
+	CloneURL      string `json:"clone_url"`
+	Private       bool   `json:"private"`
+	DefaultBranch string `json:"default_branch"`
+}
+type GithubRepositories struct {
+	Repositories []GithubRepository `json:"repositories"`
+	TotalCount   int                `json:"total_count"`
+}
 
-	jwtToken, err := g.generateJWTToken()
+func (g *github) GetListRepos(page int32, installationId, userAccessToken string) (*GithubRepositories, error) {
+	query := ""
+	if page != 0 {
+		query = fmt.Sprintf("page=%d", page)
+	}
+	endpoint := fmt.Sprintf("/user/installations/%v/repositories", installationId)
+	url := getUrl(BASE_API_URL, endpoint, query)
+
+	body, err :=
+		g.callGithub(url, "GET", genAuthToken(userAccessToken), true, nil)
 	if err != nil {
-		return "", err
+		return nil, fmt.Errorf("Error getting repo list from github. %v", err)
 	}
 
-	body, err := g.CallGithub(endpoint, "POST", "", g.GenBearer(jwtToken))
+	githubRepos := GithubRepositories{}
+
+	err = json.Unmarshal(body, &githubRepos)
 	if err != nil {
-		return "", err
+		return nil, fmt.Errorf("Error parsing github response. %v", err)
 	}
+	return &githubRepos, nil
+}
+
+type GithubUserInfo struct {
+	Email    string `json:"email"`
+	Username string `json:"login"`
+}
 
-	type installationTokenModel struct {
-		Token string `json:"token"`
+func (g *github) GetUserInformation(userAccessToken string) (*GithubUserInfo, error) {
+	url := getUrl(BASE_API_URL, "/user", "")
+	body, err :=
+		g.callGithub(url, "GET", genAuthToken(userAccessToken), true, nil)
+	bodyStr := string(body)
```

Comment:
action - this must come after the error check

Reply:
actually no it doesn't, i tested even if `bodyStr` is empty it won't error out
The reason why I am putting it before the error check is coz i am using it in multiple error cases, do u prefer something else?

Reply:
Don't u feel it's weird?
Like the function u call has an error but u are still processing the result?

Reply:
Yea man i agree will update it

Reply:
I was checking it, most of the times GitHub returns an error as text
The reason this will work, even if GitHub returns an empty byte array, I will just print no response 

Note: printing error is not enough, it just has the status code and some minor info, i refactored a bit of the code, still left the line as is, just updated the rest to follow a similar pattern (u can check the last commit)